add Quantizer support to IValue

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42619 update DispatchKey::toString()
* #42610 suppress XLAPreAutograd in AutoNonVariableTypeMode
* **#42438 add Quantizer support to IValue**

This PR adds IValue support for the `Quantizer` class, by adding a `QuantizerType` tag and associated constructors and destructors. 

Notes:
1. This eliminates the final type unsupported by our boxed kernel wrappers, so here we also take the opportunity to eliminate the artifacts of our previous incomplete support: a) the special profiling entry points that would box a "can't box" string for unsupported types, and b) the metaprogramming to generate an error-raising wrapper for kernels with `Quantizer` in their signatures.

2. `ConstQuantizerPtr` still has some special logic associated with it in the JIT code, commented as temporary; this PR leaves that undisturbed, but should facilitate its elimination in a followup.

Differential Revision: [D22894190](https://our.internmc.facebook.com/intern/diff/D22894190)

Commit:
add Quantizer support to IValue

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -205,17 +153,21 @@ struct BoxedKernelWrapper<
+  static torch::jit::Stack boxArgs(Args... args) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
+    stack.reserve(sizeof...(Args));
```
Should this reserve be called inside torch::jit::push? Maybe other call sites would benefit from it too.

bhosmer replied:
True, I can't think of a downside... OTOH I'm not sure I want to colocate that change with the other stuff in this PR 😁

Commit:
Update on "add Quantizer support to IValue"


This PR adds IValue support for the `Quantizer` class, by adding a `QuantizerType` tag and associated constructors and destructors. 

Notes:
1. This eliminates the final type unsupported by our boxed kernel wrappers, so here we also take the opportunity to eliminate the artifacts of our previous incomplete support: a) the special profiling entry points that would box a "can't box" string for unsupported types, and b) the metaprogramming to generate an error-raising wrapper for kernels with `Quantizer` in their signatures.

2. `ConstQuantizerPtr` still has some special logic associated with it in the JIT code, commented as temporary; this PR leaves that undisturbed, but should facilitate its elimination in a followup.

Differential Revision: [D22894190](https://our.internmc.facebook.com/intern/diff/D22894190)

[ghstack-poisoned]

Commit:
Update on "add Quantizer support to IValue"


This PR adds IValue support for the `Quantizer` class, by adding a `QuantizerType` tag and associated constructors and destructors. 

Notes:
1. This eliminates the final type unsupported by our boxed kernel wrappers, so here we also take the opportunity to eliminate the artifacts of our previous incomplete support: a) the special profiling entry points that would box a "can't box" string for unsupported types, and b) the metaprogramming to generate an error-raising wrapper for kernels with `Quantizer` in their signatures.

2. `ConstQuantizerPtr` still has some special logic associated with it in the JIT code, commented as temporary; this PR leaves that undisturbed, but should facilitate its elimination in a followup.

Differential Revision: [D22894190](https://our.internmc.facebook.com/intern/diff/D22894190)

[ghstack-poisoned]

Commit:
Update on "add Quantizer support to IValue"


This PR adds IValue support for the `Quantizer` class, by adding a `QuantizerType` tag and associated constructors and destructors. 

Notes:
1. This eliminates the final type unsupported by our boxed kernel wrappers, so here we also take the opportunity to eliminate the artifacts of our previous incomplete support: a) the special profiling entry points that would box a "can't box" string for unsupported types, and b) the metaprogramming to generate an error-raising wrapper for kernels with `Quantizer` in their signatures.

2. `ConstQuantizerPtr` still has some special logic associated with it in the JIT code, commented as temporary; this PR leaves that undisturbed, but should facilitate its elimination in a followup.

Differential Revision: [D22894190](https://our.internmc.facebook.com/intern/diff/D22894190)

[ghstack-poisoned]
